File: app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/AbstractCreateNoteFragment.java
package de.westnordost.streetcomplete.data.osmnotes;

import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.BottomSheetBehavior;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.quests.AbstractBottomSheetFragment;
import de.westnordost.streetcomplete.quests.note_discussion.AttachPhotoFragment;
import de.westnordost.streetcomplete.util.TextChangedWatcher;

public abstract class AbstractCreateNoteFragment extends AbstractBottomSheetFragment
{
	private EditText noteInput;
	private View buttonOk;
	private TextView textTitle;
	private TextView textDescription;

	@Override public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
									   Bundle savedInstanceState)
	{
		View view = inflater.inflate(getLayoutResId(), container, false);

		LinearLayout bottomSheet = view.findViewById(R.id.bottomSheet);
		if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE)
		{
			BottomSheetBehavior.from(bottomSheet).setState(BottomSheetBehavior.STATE_EXPANDED);
		}

		textTitle = view.findViewById(R.id.title);

		ViewGroup content = view.findViewById(R.id.content);
		content.removeAllViews();
		inflater.inflate(R.layout.form_leave_note, content);

		textDescription = content.findViewById(R.id.description);

		noteInput = content.findViewById(R.id.noteInput);
		noteInput.addTextChangedListener(new TextChangedWatcher(this::updateOkButtonEnablement));

		ViewGroup buttonPanel = view.findViewById(R.id.buttonPanel);
		buttonPanel.removeAllViews();
		inflater.inflate(R.layout.quest_buttonpanel_ok_cancel, buttonPanel);

		buttonPanel.findViewById(R.id.buttonCancel).setOnClickListener(v -> getActivity().onBackPressed());
		buttonOk = buttonPanel.findViewById(R.id.buttonOk);
		buttonOk.setOnClickListener(v -> onClickOk());

		updateOkButtonEnablement();

		return view;
	}

	public void setTitle(@StringRes int resId) { textTitle.setText(resId); }
	public void setDescription(@StringRes int resId) { textDescription.setText(resId); }

	@Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState)
	{
		super.onViewCreated(view, savedInstanceState);
		if(savedInstanceState == null)
		{
			getChildFragmentManager().beginTransaction().add(R.id.attachPhotoFragment, new AttachPhotoFragment()).commit();
		}
	}

	private @Nullable AttachPhotoFragment getAttachPhotoFragment()
	{
		return (AttachPhotoFragment) getChildFragmentManager().findFragmentById(R.id.attachPhotoFragment);
	}

	private void onClickOk()
	{
		AttachPhotoFragment f = getAttachPhotoFragment();
		onLeaveNote(getNoteText(), f != null ? f.getImagePaths() : null);
	}

	@Override protected void onDiscard()
	{
		AttachPhotoFragment f = getAttachPhotoFragment();
		if(f != null) f.deleteImages();
	}

	@Override public boolean isRejectingClose()
	{
		return !getNoteText().isEmpty();
	}

	private String getNoteText()
	{
		return noteInput.getText().toString().trim();
	}

	private void updateOkButtonEnablement()
	{
		buttonOk.setEnabled(!getNoteText().isEmpty());
	}

	protected abstract void onLeaveNote(String text, @Nullable ArrayList<String> imagePaths);
	@LayoutRes protected abstract int getLayoutResId();
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/AbstractCreateNoteFragment.kt
package de.westnordost.streetcomplete.data.osmnotes

import android.content.res.Configuration
import android.os.Bundle
import com.google.android.material.bottomsheet.BottomSheetBehavior
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.quests.AbstractBottomSheetFragment
import de.westnordost.streetcomplete.quests.note_discussion.AttachPhotoFragment
import de.westnordost.streetcomplete.util.TextChangedWatcher
import kotlinx.android.synthetic.main.form_leave_note.*
import kotlinx.android.synthetic.main.quest_buttonpanel_done_cancel.*

abstract class AbstractCreateNoteFragment : AbstractBottomSheetFragment() {

    private val attachPhotoFragment: AttachPhotoFragment?
        get() = childFragmentManager.findFragmentById(R.id.attachPhotoFragment) as AttachPhotoFragment

    private val noteText get() = noteInput?.text?.toString().orEmpty().trim()

    protected abstract val layoutResId: Int

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(layoutResId, container, false)

        val bottomSheet = view.findViewById<LinearLayout>(R.id.bottomSheet)
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            BottomSheetBehavior.from(bottomSheet).state = BottomSheetBehavior.STATE_EXPANDED
        }

        val content = view.findViewById<ViewGroup>(R.id.content)
        content.removeAllViews()
        inflater.inflate(R.layout.form_leave_note, content)

        val buttonPanel = view.findViewById<ViewGroup>(R.id.buttonPanel)
        buttonPanel.removeAllViews()
        inflater.inflate(R.layout.quest_buttonpanel_done_cancel, buttonPanel)

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState == null) {
            childFragmentManager.beginTransaction()
                .add(R.id.attachPhotoFragment, AttachPhotoFragment())
                .commit()
        }

        noteInput.addTextChangedListener(TextChangedWatcher { updateDoneButtonEnablement() })

        cancelButton.setOnClickListener { activity!!.onBackPressed() }
        doneButton.setOnClickListener { onClickOk() }

        updateDoneButtonEnablement()
    }

    private fun onClickOk() {
        onComposedNote(noteText, attachPhotoFragment?.imagePaths)
    }

    override fun onDiscard() {
        attachPhotoFragment?.deleteImages()
    }

    override fun isRejectingClose() =
        noteText.isNotEmpty() || attachPhotoFragment?.imagePaths?.isNotEmpty() == true

    private fun updateDoneButtonEnablement() {
        doneButton.isEnabled = !noteText.isEmpty()
    }

    protected abstract fun onComposedNote(text: String, imagePaths: List<String>?)
}
